accepted/2022/simplify-rid-model.md

```diff
@@ -0,0 +1,252 @@
+- Implement a new algorithmic host scheme for RID selection, enabled by default.
+
+The new scheme will be based on the following:
+
+- Each host is built with a hard-coded set of RIDs that it probes for, both triplets and singles.
+- Each host is already built for a specific RID triple, so this change is an evolutionary change.
+- There are processor-agnostic managed-code scenarios where a RID single is relevant, like Windows-only code (for example, accessing the registry) that works the same x64, Arm64 and x86. The same is true for macOS and Linux.
+- The RID that `dotnet --info` returns will always match the first in this hard-coded set of RIDs.
+
+Let's assume that an app is running on Alpine 3.16, the host will probe for the following RIDs:
+
+- `linux-musl-x64`
+- `unix`
+
+A glibc distro like Debian would be similar:
+
+- `linux-arm64`
+- `unix`
+
+macOS would be similar:
+
+- `osx-arm64`
+- `unix`
+
+Note: The abstract `unix` RID is used for macOS and Linux, instead of a concrete RID for those OSes.
+
+Note: We may want to add `osx` to describe macOS-specific processor-agnostic managed code. If we find some scenarios that need it, we should add it. 
+
+Windows would similar:
+
+- `win-x86`
+- `win`
+
+Note: `win7` and `win10` also exist (as processor-specific and -agnostic). Ideally, we don't have to support those in the host-specific scheme, but we need to do some research on that.
+
+Note: A mix of processor-specific RIDs are used above, purely to demonstrate the range of processor-types supported. Each operating system supports a variety of processor-types, and this is codified in their associated RIDs.
+
+More generally, the host will probe for:
+
+- This environment, RID triplet (OS-CRuntime-arch)
+- This environment, RID single (OS-only).
+
+The host will implement "first one wins" probing logic for selecting a RID-specific asset.
+
+This behavior only applies for portable apps. RID-specific apps will not probe for RID-specific assets.
+
+There are other RIDs in `runtimes.json` today. This new scheme will not support those. The `runtimes.json` host compat feature will need to be used to support those, which is likely very uncommon. Even if they are quasi-common, we will likely still make this change and require some combination of the ecosystem to adapt and app developers to opt-in to the `runtimes.json` compat mode.
+
+The host and `runtimes.json` must remain compatible. The RIDs known by the host must be a subset of `runtimes.json` RIDs. We may need to update `runtimes.json` to ensure that the RIDs known by the host are present in that file.
+
+
+## Minimum CRT/libc version
+
+This scheme doesn't explicitly define a CRT/libc version. Instead, we will define a minimum CRT/libc version, per major .NET version. That minimum CRT/libc version will form a contract and should be documented, per .NET version. Practically, it defines the oldest OS version that you can run an app on, for a given .NET version.
```

Comment:
Using C runtime version to define the OS version makes sense on Linux. It is the best approximation of the OS version to use for determining binary compatibility (of binaries written in C/C++).

Using C runtime version for this purpose makes less sense on Windows and macOS. Windows and macOS have well-defined OS version. Both Windows (Windows SDK) and macOS (XCode) native toolchains are oriented on using the OS version for this purpose ([WINVER](https://docs.microsoft.com/en-us/cpp/porting/modifying-winver-and-win32-winnt) on Windows and `-mmacos-version-min` on macOS). We should use OS version for determining binary compatibility on Windows and macOS.

Reply:
Do you want me to add that text, or adapt its meaning into the current text?

Reply:
I think adapting the meaning into the current text would look better. I would lead with minimum OS version as a documented contract, per .NET version. libc version is Linux-specific approximation of OS version across distros for the purposes of this contract.